File: B10.cpp
#include <string>
#include <iostream>
#include <iomanip>

using namespace std;
// for writing function - body is :
// return_type function_name( parameter list ) {
  // body of the function
// }

// auxiliary function for checking leap year. //The return type is bool, which means that every return statement
//has to provide a bool expression. The first line outputs the value true because 2 is a single-digit number.
//does not display the words true and false, but rather the integers 1 and 0.
bool leapyear(int year) {
    if (year % 4 == 0 && year % 100 != 0 || year % 400 == 0)
        return true;
    return false;
}
// get weekday from entered date - Sakamoto algorythm
int dayofweek(int day, int month, int year){
    static int t[] = { 0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4 };
    year -= month < 3;
    return ( year + year/4 - year/100 + year/400 + t[month-1] + day) % 7;

}
// Do the black job
int main()
{
    char fill = '/';
    string entered_date;
    int entered_month;
    int entered_year;
    int entered_day;

    cout << "Enter a date (dd/mm/yyyy): " << endl;
    cin >> entered_date;
    //parse integers out of strings with stoi. 1st - order, 2nd - length. this one I found really convenient
    entered_month = stoi(entered_date.substr(3,2));
    entered_day = stoi(entered_date.substr(0,2));
    entered_year = stoi(entered_date.substr(6));
        // check entered values
    //cout << entered_year << entered_month << entered_day << endl;

// validates entered date
// make dictionary
    const int lookup_table[12] = {31,29,31,30,31,30,31,31,30,31,30,31};
    if (!(entered_month >= 1 && entered_month <= 12)) {
        cout<<"wrong month";
        return 1;
    }
    if (!(entered_day >= 1 && entered_day <= lookup_table[entered_month-1])) {
        cout<<"wrong day";
        return 1;
    }
    if (!(entered_year >= 0001)) {  // 0001 format cannot be used in python
        cout<<"wrong year";
        return 1;
    }

    // Call dayofweek function ; While creating a C++ function, you give a definition of what the function has to do.
    // To use a function, you will have to call or invoke that function.

int day_num = dayofweek(entered_day, entered_month, entered_year);
const string day_names[] = {"Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"};
cout<<day_names[day_num]<<endl;



    // Call leapyear function
    if (leapyear(entered_year))
        cout<<entered_year<<" is a leap year";
    else
        cout<<entered_year<<" is not a leap year";
    return 0;
}

File: B10.py
# Create program both in C++, and Python by making use of described or other functions.
# #Requirements are the same as for the previous task, see published detailed requirements.
# TASK B10. Given three natural numbers which denote date (day, month, year).
# #Create function which returns weekday for that date. - > DONE
# The solution should also contain an auxiliary  function for leap year checking. -> DONE
### easiest ways - date or calendar libraries; in this case no built in libraries are allowed, so
##Zeller’s formula can be used
##############################INPUT################################################
date_str = input('Enter some date in exactly following format: dd mm yyyy \n')
day = int(date_str.split(' ')[0])
month = int(date_str.split(' ')[1])
year = int(date_str.split(' ')[2])

###########################VALIDATES INPUT##################################

if month > 0 and month <= 12:
    print("month ok")
else:
    print("month NOT ok")
##### use dictionaries for day check
lookup_table = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
if not (day >= 1 and day <= lookup_table[month - 1]):
    print("wrong day \n")


## check length of the year
if 1 <= len(str(year)) <= 4:
    print("year ok")
else:
    print("year NOT ok")


#############################DO THE BLACK WORK ##################################################
def day_of_week(year, month, day):
    if (month < 3):
        month += 12
        year -= 1
        # Sakamoto's algorithm to find out the day of week from a given date.
    result = (day + (int((13 * (month - 2)) / 5)) + year + (int(year / 4)) - (int(year / 100)) - (int(year / 400))) % 7
    return int(result) + 1
### make array with week name
week_days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
week_num = day_of_week(year, month, day)
print(week_days[week_num])

# function for determining leap year; source logics from wiki: https://en.wikipedia.org/wiki/Leap_year_problem

def leap_year(year):
    if year % 4 == 0:
        if year % 100 != 0:
            return True
    if year % 400 == 0:
        return True
    return False

if leap_year(year):
    # The format() method allows you format string in any way you want.
    # {0} and {1} are format codes. The format code {0} is replaced by the first argument of
    # format() i.e 12 , while {1} is replaced by the second argument of format() i.e 31 .
    print("{0} is a leap year".format(year))
else:
    print("{0} is not a leap year".format(year))

    #######################LAZY RESERVE SOLUTION DO NOT TAKE THIS INTO ACCOUNT ################################
# import datetime
#
# def day_of_week(date):
#     day = date.weekday()
#     # make array with day names
#     day_name = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
#     return day_name[day]
#
# date_str = input('Enter some date in following format: dd mm yyyy \n')
# # The strptime() method creates a datetime object from the given string.
# #Note: You cannot create datetime object from every string. The string needs to be in a certain format.
# date = datetime.datetime.strptime(date_str, '%d %m %Y')
# print(day_of_week(date))
